File: analyse_Narasimhan_fig2/finalize_fig2B.py
#!/usr/bin/env python3

import sys
import random

def get_data(neutral, deleterious, pops) :
    print('Reading data...', file=sys.stderr)
    if 'CADD' in neutral :
        filename = 'counts_fig2B_CADD.txt'
    else :
        filename = 'counts_fig2B.txt'
    f = open(filename)
    next(f)                     # header
    dico = {}
    for line in f :
        chrom, pos, annot, pop, hom, alt = line.split()
        if annot != neutral and annot != deleterious :
            continue
        if (chrom, pos) not in dico :
            dico[(chrom, pos)] = {'annot':annot}
        dico[(chrom,pos)][pop] = (int(hom), int(alt))
        
    f.close()
    return dico

def count_hom_deleterious(data, deleterious, pops) :
    nb_hom = {}
    nb_alt = {}
    for pop in pops :
        nb_hom[pop] = 0
        nb_alt[pop] = {}
    for coords in data :
        annot = data[coords]['annot']
        if annot != deleterious :
            continue
        for pop in pops :
            tmp_hom, tmp_alt = data[coords][pop]
            if tmp_hom > 0 :
                nb_hom[pop] += 1
            if tmp_alt != 0 :
                if tmp_alt not in nb_alt[pop] :
                    nb_alt[pop][tmp_alt] = 0
                nb_alt[pop][tmp_alt] += 1
    return nb_hom, nb_alt

def prepare_neutral_data(data, pops, neutral) :
    dico = {}
    for pop in pops :
        dico[pop] = {}
    for coords in data :
        annot = data[coords]['annot']
        if annot != neutral :
            continue
        for pop in pops :
            tmp_hom, tmp_alt = data[coords][pop]
            if tmp_hom > 0 :
                tmp_hom = 1
            if tmp_alt != 0 :
                if tmp_alt not in dico[pop] :
                    dico[pop][tmp_alt] = []
                dico[pop][tmp_alt].append(tmp_hom)
    return dico

def perform_resampling(dico_hom, dico_freqs, pops) :
    result = {}
    for pop in pops :
        result[pop] = 0
        for freq in dico_freqs[pop] :
            count = dico_freqs[pop][freq]
            tmp = random.sample(dico_hom[pop][freq], count)
            result[pop] += sum(tmp)
    return result

def main() :
    pops = ['CTA','LPO','TUR','TJE']
    nb_resampling = 1000
    deleterious = 'CADD_sup_28'
    neutral = 'CADD_inf_10'
    data = get_data(neutral, deleterious, pops)
    nb_hom_deleterious, alt_freq_deleterious = count_hom_deleterious(data, deleterious, pops)
    dico_hom_neutral = prepare_neutral_data(data, pops, neutral)
    new_line = ['idx'] + pops
    print('\t'.join(new_line))
    new_line = ['0']
    for pop in pops :
        new_line.append(str(nb_hom_deleterious[pop]))
    print('\t'.join(new_line))
    for idx in range(nb_resampling) :
        new_line = [f'{idx+1}']
        resampling = perform_resampling(dico_hom_neutral, alt_freq_deleterious, pops)
        for pop in pops :
            new_line.append(str(resampling[pop]))
        print('\t'.join(new_line))
    
    
if __name__ == '__main__' :
    main()

File: analyse_Narasimhan_fig2/finalize_fig2B.R

library(vioplot)

if (!('d' %in% ls())){
    d <- read.table('counts_fig2B.txt', header=TRUE)
}

nb.simu <- 1000

sample.sizes <- c(22,21,24,19)
names(sample.sizes) <- c('CTA','LPO','TUR','TJE')
maf <- 0.1

pdf(sprintf('fig2B_MAF_%.2f.pdf', maf),  10, 7, title='Expected homozygote LOF variants')
par(mfrow=c(1,4))


pops <- unique(d$pop)
for (pop in pops){
    cat(pop,'\n')
    tmp.d <- d[d$pop==pop,]
    max.nb.alt <- maf * (2*sample.sizes[pop])
    tmp.d <- tmp.d[tmp.d$alt <= max.nb.alt,] 
    
    syns <- tmp.d[tmp.d$annot=='VEP_SS',]
    lof <- tmp.d[tmp.d$annot=='VEP_LOF',]
    expected <- c()
    ## we extract the frequencies (# alt allele) of LOF variants
    counts.nb.alt <- table(lof$alt)
    freqs <- as.numeric(names(counts.nb.alt))
    ## for each simu
    for (i in 1:nb.simu){
        random <- c()
        for (j in 1:length(counts.nb.alt)){ # for each frequency class
            tmp.freq <- freqs[j]
            nb.wanted <- counts.nb.alt[j]
            ## we identify the SYN variants with same frequency
            to.sample.from <- syns[syns$alt==tmp.freq,]
            ## we randomly draw from them
            tmp <- sample(to.sample.from$hom, nb.wanted)
            ## aggregate result
            random <- c(random, length(which(tmp > 0)))
        }
        expected <- c(expected, sum(random))
    }
    observed <- length(which(lof$hom > 0))
    vioplot(expected, ylim=range(c(expected, observed)), main=sprintf('%s MAF %.1f', pop, maf), xlab='', ylab='# variants with homozygote genotypes',xaxt='n')
    quantiles <- quantile(expected, probs=c(.025,.975))
    abline(h=quantiles,lty=2)
    points(1, observed, col='red',pch=19)
}

## max.freq <- .1
## for (pop in pops){
##     cat(pop,'\n')
##     tmp <- system(sprintf('wc -l ../samples_%s.txt', pop), intern=TRUE)
##     nb.sample <- as.integer(strsplit(tmp,' ')[[1]][1])
##     tmp.d <- d[d$pop==pop,]
##     syns <- tmp.d[tmp.d$annot=='VEP_SS',]
##     lof <- tmp.d[tmp.d$annot=='VEP_LOF',]
##     expected <- c()
##     ## we extract the frequencies (# alt allele) of LOF variants
##     counts.nb.alt <- table(lof$alt)
##     freqs <- as.numeric(names(counts.nb.alt))
##     ## only keep freqs less than what we want
##     to.keep <- freqs / (2*nb.sample) < max.freq
##     counts.nb.alt <- counts.nb.alt[to.keep]
##     freqs <- as.numeric(names(counts.nb.alt))
##     ## for each simu
##     for (i in 1:nb.simu){
##         random <- c()
##         for (j in 1:length(counts.nb.alt)){ # for each frequency class
##             tmp.freq <- freqs[j]
##             nb.wanted <- counts.nb.alt[j]
##             ## we identify the SYN variants with same frequency
##             to.sample.from <- syns[syns$alt==tmp.freq,]
##             ## we randomly draw from them
##             tmp <- sample(to.sample.from$hom, nb.wanted)
##             ## aggregate result
##             random <- c(random, length(which(tmp > 0)))
##         }
##         expected <- c(expected, sum(random))
##     }
##     ## compute statistic we are interested in
##     freqs <- lof$alt / (2*nb.sample)
##     to.keep <- freqs < max.freq
##     observed <- length(which(lof$hom[to.keep] > 0))

##     ## plot result
##     vioplot(expected, ylim=range(c(expected, observed)), main=pop, xlab='', ylab='# variants with homozygote genotypes',xaxt='n')
##     quantiles <- quantile(expected, probs=c(.025,.975))
##     abline(h=quantiles,lty=2)
##     points(1, observed, col='red',pch=19)
## }


dev.off()
